### PR TITLE
Case Dismissed: Forcibly show DAppConnectionInfoBar popover on first dApp connection

### PR DESCRIPTION
**Pull request by Shadowfiend**
Draft until #3451 lands.

Adds a `DismissableItem` concept and related functionality to the
`PreferencesService` and to the UI slice, providing a common pattern
for things that should be shown to the user exactly once, e.g. new
feature banners and related functions.

This PR will complete the work on updated dApp connections.

Latest build: [extension-builds-3464](https://github.com/tahowallet/extension/suites/13641442570/artifacts/752912241) (as of Thu, 15 Jun 2023 20:55:13 GMT).

### CODE REVIEW

File: background/services/preferences/db.ts [hunk 1]
```diff
@@ -36,6 +36,29 @@ export interface Preferences {
   }
 }
 
+/**
+ * Items that the user will see and then will not reappear once they've been
+ * manually dismissed.
```
Comment (Shadowfiend) [inline background/services/preferences/db.ts:41]: This can also be a “Don't show again”-style interaction, I'll add a note to clarify that.

File: background/main.ts [hunk 2]
```diff
@@ -1558,6 +1575,8 @@ export default class Main extends BaseService<never> {
           newDefaultWalletValue
         )
 
+        // FIXME Both of these should be done as observations of the preference
+        // FIXME service event rather than being managed by `main`.
```
Comment (hyphenized) [inline background/main.ts:1579]: This pattern seems great but I reckon the logic is going to get a bit hard to follow 🤔. Do we want to only execute calls to other services on events emitted by services as reactions to mutations?

**Review by hyphenized — COMMENTED** [review]
Tested and looks good! Left a few comments

File: background/main.ts (hunk 2 above)
Comment (Shadowfiend) [reply background/main.ts:1579]: Definitely yes for cross-service stuff. No reason Main should be deciding what the ProviderBridgeService should be doing with an event—that is the ProviderBridgeService's responsibility. The only such triggers from Main should be directly due to redux actions, and in any such action we should be asking “is the service reacting to the user's change, or is the service reacting to a system change”. In this case, the preferences service reacts to the user's change, and that system then becomes a system change that triggers other adjustments. This is just separation of responsibilities, but we haven't done a great job of following it everywhere.

The short articulation is: services are responsible for their own state and for reacting to the state of other services. This is another reason why Main really needs to be renamed to ReduxService; it's charged with mediating with redux, it is NOT some sort of superservice meant to mediate between many services, unless there's some redux-based thing that needs to intervene. Similarly, this is why settings need to make it into PreferencesService—that allows it to be the source of truth for settings, not the redux store, which should generally not be the source of truth for anything.

This opens up the possibility of further paring down the redux store by only storing “recent” things in it, and fetching async from the backend for less recent things (e.g. transactions). I think that's months away, and it won't make sense for everything, but it's what the overall architectural approach is rooted in.

File: background/services/preferences/db.ts [hunk 3]
```diff
@@ -36,6 +36,30 @@ export interface Preferences {
...
+ * manually dismissed. Manual dismissal can include closing a popover, or
+ * selecting "Don't show again" on a popup before closing it.
+ */
+export type ManuallyDismissableItem = "analytics-enabled-banner"
+/**
+ * Items that the user will see once and will not be auto-displayed again. Can
+ * be used for tours, or for popups that can be retriggered but will not
+ * auto-display more than once.
+ */
+export type SingleShotItem = "default-connection-popover"
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply background/services/preferences/db.ts:50]: So the distinction I'm trying to draw here, and I think I've done a pretty poor job of it lol, is between items that are dismissed through a user's “please don't show this to me again”-type action, or a banner that needs to be explicitly closed but doesn't block content, and items that are dismissed automatically. I think `SingleShotItem` is actually a decent name for the latter, but `DismissableItem` as the umbrella type doesn't reflect that very well.

Maybe it's just `AutomaticallyDismissableItem` as a counterpart to `ManuallyDismissableItem` though? 🤔

File: ui/components/DAppConnection/DAppConnectionInfoBar.tsx [hunk 4]
```diff
@@ -24,7 +29,7 @@ function DefaultConnectionPopover({ close }: PopoverProps): ReactElement {
   return (
     <div
       className={classNames("bg", {
-        fadeIn: !isClosing,
+        fade_in: !isClosing,
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply ui/components/DAppConnection/DAppConnectionInfoBar.tsx:32]: Oiiiii how did it sneak innnnnn.

Since this is in the new stuff and I have these changes, how about we keep it and I open the follow-up PR that fixes the `fade_in`?

File: ui/components/DAppConnection/DAppConnectionInfoBar.tsx [hunk 5]
```diff
@@ -176,34 +181,54 @@ function DefaultConnectionPopover({ close }: PopoverProps): ReactElement {
...
+    selectShouldShowDismissableItem("default-connection-popover")
+  )
+
+  const [
+    isShowingDefaultConnectionPopover,
+    setIsShowingDefaultConnectionPopover,
+  ] = useState(shouldImmediatelyShowPopover)
+
+  const closeDefaultConnectionPopover = useCallback(() => {
+    if (shouldImmediatelyShowPopover && isShowingDefaultConnectionPopover) {
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply ui/components/DAppConnection/DAppConnectionInfoBar.tsx:196]: I have to tell you that my radar for when `useCallback` is the right move and when it's the wrong one is as broken as the tree that was used to make my desk.

[the remark this answers is not in the document]
Comment (Shadowfiend) [reply ui/components/DAppConnection/DAppConnectionInfoBar.tsx:196]: Sorry, I missed the first part of the note—why do you think that? It seems like the correct condition, no? If it's not showing and we're set to immediately show it, we shouldn't dispatch anything, because that means it's closing.

File: ui/components/DAppConnection/DAppConnectionInfoBar.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply ui/components/DAppConnection/DAppConnectionInfoBar.tsx:32]: Went ahead and dropped this into https://github.com/tahowallet/extension/pull/3485 .

File: background/services/preferences/db.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (hyphenized) [reply background/services/preferences/db.ts:50]: Ok then I think this is fine as it is 👍

[the remark this answers is not in the document]
Comment (Shadowfiend) [reply background/services/preferences/db.ts:50]: Happy to see/hear ideas for a better name here. Fortunately it's in the type, so it doesn't get persisted, so we can deliberate a little without blocking the merge.

File: ui/components/DAppConnection/DAppConnectionInfoBar.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (hyphenized) [reply ui/components/DAppConnection/DAppConnectionInfoBar.tsx:196]: Because this only runs when the popover is open right?

File: background/services/preferences/db.ts [hunk 6]
```diff
@@ -319,6 +340,12 @@ export class PreferenceDatabase extends Dexie {
         })
     })
 
+    this.version(17).stores({
+      preferences: "++id",
+      signersSettings: "&id",
+      shownDismissableItems: "&id,shown",
```
Comment (kkosiorowska) [inline background/services/preferences/db.ts:346]: Boolean can't be indexed, so adding a `shown` here does not make much sense.  

https://dexie.org/docs/Version/Version.stores()

File: ui/components/DAppConnection/DAppConnectionInfoBar.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply ui/components/DAppConnection/DAppConnectionInfoBar.tsx:196]: I always assume propagation delays for stuff like this, not least because there are cases where we delay rendering changes to wait on animations, etc. We can always case by case those, but it's easier to just consistently get the condition right so reasoning is localized rather than requiring global knowledge of caller behavior.

File: background/services/preferences/db.ts (hunk 6 above)
Comment (Shadowfiend) [reply background/services/preferences/db.ts:346]: Bah, yeah. I'd rather not add a migration here but maybe it won't be necessary given we haven't shipped this yet.
